src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnectionImpl.java

```diff
@@ -1274,19 +1224,11 @@ public NotificationResult fetchNotifications(long clientSequenceNumber,
                                        + "returns null to force the client to stop fetching");
                 return null;
             }
-            final long csn = clientSequenceNumber;
-            final int mn = maxNotifications;
-            final long t = timeout;
-            PrivilegedAction<NotificationResult> action =
-                new PrivilegedAction<NotificationResult>() {
-                    public NotificationResult run() {
-                        return getServerNotifFwd().fetchNotifs(csn, t, mn);
-                    }
-            };
+
             if (subject == null) {
-                return action.run();
+                return getServerNotifFwd().fetchNotifs(clientSequenceNumber, timeout, maxNotifications);
             } else {
-                return Subject.doAs(subject, action);
+                return Subject.callAs(subject, () -> getServerNotifFwd().fetchNotifs(clientSequenceNumber, timeout, maxNotifications));
```

Comment:
This call may throw CompletionException instead of PrivilegedActionException
That's strange that this case does not go through standard `doPrivilegedOperation` way.
I think this class needs to be improved (most likely as a separate PR).
For most of operations it created "operation id" and creates array of arguments, but then `doOperation` parses the id and arguments and performs required action. It would be much simpler if doPrivilegedOperation/doOperation accept Callable.

Reply:
Yes thanks, can unwrap CompletionException here like in the other callAs() usage in doPrivilegedOperation().